Download client for qBittorrent

cbodley opened the pull request:
I had experimented with a qBittorrent plugin a few months ago, but I wasn't able to get it working well because their Web UI API (https://github.com/qbittorrent/qBittorrent/wiki/WebUI-API-Documentation) didn't have a way to set labels. I've been tracking their "[Web UI] Labels implementation" pull request at https://github.com/qbittorrent/qBittorrent/pull/3444, and it looks like it may merge in the near future.

I updated my prototype to make use of the setLabel command it introduces, so I wanted to share it with you ahead of time for feedback.

markus101 commented on src/NzbDrone.Core/Download/Clients/qBittorrent/QBittorrent.cs:
```diff
@@ -0,0 +1,228 @@
+                        item.Status = DownloadItemStatus.Paused;
+                        break;
+
+                    case "queuedUP": // queuing is enabled and torrent is queued for upload
```
This should be completed.

markus101 commented on src/NzbDrone.Core/Download/Clients/qBittorrent/QBittorrent.cs:
```diff
@@ -0,0 +1,228 @@
+
+                    case "queuedUP": // queuing is enabled and torrent is queued for upload
+                    case "queuedDL": // queuing is enabled and torrent is queued for download
+                    case "checkingUP": // torrent has finished downloading and is being checked
```
Completed as well.

markus101 commented on src/NzbDrone.Core/Download/Clients/qBittorrent/QBittorrentProxy.cs:
```diff
@@ -0,0 +1,187 @@
+
+        public void RemoveTorrent(String hash, Boolean removeData, QBittorrentSettings settings)
+        {
+            String cmd = removeData ? "/command/deletePerm" : "/command/delete";
```
`var` instead of `String`

markus101 commented on src/NzbDrone.Core/Download/Clients/qBittorrent/QBittorrentTorrent.cs:
```diff
@@ -0,0 +1,24 @@
+{
+    public class QBittorrentTorrent
+    {
+        public String Hash { get; set; } // Torrent hash
```
Things aren't currently strict, but we prefer, string, int, long and bool instead of String, Int32, Int64 and Boolean.

markus101 replied:
I won't comment everywhere since there are a number of cases, but please update those as well.

markus101 commented on src/NzbDrone.Core/Download/Clients/qBittorrent/QBittorrentSettings.cs:
```diff
@@ -0,0 +1,56 @@
+        }
+
+        [FieldDefinition(0, Label = "Host", Type = FieldType.Textbox)]
+        public String Host { get; set; }
```
Same here, `string` instead of `String`.

markus101 commented on src/NzbDrone.Core/Download/Clients/qBittorrent/QBittorrentProxy.cs:
```diff
@@ -0,0 +1,187 @@
+        }
+    }
+
+    public class DigestAuthenticator : IAuthenticator
```
This should be in its own file.

markus101 commented on src/NzbDrone.Core/Download/Clients/qBittorrent/QBittorrentProxy.cs:
```diff
@@ -0,0 +1,187 @@
+            {
+                _logger.Debug("QBittorrent authentication error.");
+
+                Login(client, settings);
```
This will trigger a login every request, instead we can cache that we authenticated correctly (same as we do for Plex Server).

markus101 commented on src/NzbDrone.Core/Download/Clients/qBittorrent/QBittorrent.cs:
```diff
@@ -0,0 +1,228 @@
+        {
+            var config = _proxy.GetConfig(Settings);
+
+            OsPath destDir = new OsPath((String)config.GetValueOrDefault("save_path"));
```
`var` instead of `OsPath`

markus101 commented on src/NzbDrone.Core/Download/Clients/qBittorrent/QBittorrent.cs:
```diff
@@ -0,0 +1,228 @@
+                item.DownloadClient = Definition.Name;
+                item.RemainingSize = (long)(torrent.Size * (1.0 - torrent.Progress));
+                if (torrent.Eta != -1)
+                    item.RemainingTime = TimeSpan.FromSeconds(torrent.Eta);
```
```
if (torrent.Eta != -1)
{
    item.RemainingTime = TimeSpan.FromSeconds(torrent.Eta);
}
```

and a line break before the if starts.

markus101 commented on src/NzbDrone.Core/Download/Clients/qBittorrent/QBittorrent.cs:
```diff
@@ -0,0 +1,228 @@
+                var outputPath = _remotePathMappingService.RemapRemoteToLocal(Settings.Host, status.OutputRootFolders[0]);
+
+                if (outputPath == null || outputPath.FileName == torrent.Name)
+                    item.OutputPath = outputPath;
```
Braces around the if and else.

markus101 commented on src/NzbDrone.Core/Download/Clients/qBittorrent/QBittorrent.cs:
```diff
@@ -0,0 +1,228 @@
+                if (torrent.Eta != -1)
+                    item.RemainingTime = TimeSpan.FromSeconds(torrent.Eta);
+
+                var outputPath = _remotePathMappingService.RemapRemoteToLocal(Settings.Host, status.OutputRootFolders[0]);
```
What if the download is under a different root folder?

markus101 commented on src/NzbDrone.Core/Download/Clients/qBittorrent/QBittorrentPriority.cs:
```diff
@@ -0,0 +1,8 @@
+        Last = 0,
+        First = 1
+    }
+}
```
nit: newline at the end of the file.

markus101 commented on src/NzbDrone.Core/Download/Clients/qBittorrent/QBittorrent.cs:
```diff
@@ -0,0 +1,228 @@
+        private readonly IQBittorrentProxy _proxy;
+
+        public QBittorrent(IQBittorrentProxy proxy,
+                        ITorrentFileInfoReader torrentFileInfoReader,
```
Align classes to the first line.

markus101 commented on src/NzbDrone.Core/Download/Clients/qBittorrent/QBittorrent.cs:
```diff
@@ -0,0 +1,228 @@
+                        break;
+
+                    default:
+                        throw new Exception("Invalid torrent state " + torrent.State);
```
If a new state is added this will break the integration completely since each torrent isn't handled in a try/catch block. We either need to wrap each torrent processing in a try/catch and log the error or log an error here and default the state to something, downloading maybe? Its kind of a toss up which one is best, I'll let the other devs weigh in here.

cbodley commented:
@markus101, thank you for all the great feedback! I don't have a lot of experience with the language, so I'm going to blame visual studio for most of the formatting stuff :) I'll gladly fix those, and put some more thought into the other issues as well.

markus101 commented:
Thanks for contributing! I know we're a bit picky with formatting, so thanks for addressing those.

I took a look at implementing qBittorrent a couple weeks back and ran into similar issues, the blocker being the lack of label changing in the API, but I didn't like how the paths are returned and ended up hitting the API again to get the General Properties for each torrent, which is ugly and expensive. Perhaps we should ask about having it returned with the initial GetTorrents request?

cbodley commented:
@markus101 Okay, now I'm remembering the issues I had with OutputPath. I had ended up just defaulting everything to the global save_path from /query/preferences.

I went ahead and implemented the call to /query/propertiesGeneral for every torrent in GetItems() to check for a save_path. And I agree that we should work with the qBittorrent devs to get the save_path included with /query/torrents, so I added a patch that checks for that before falling back to /query/propertiesGeneral.

JustMaier commented:
Hey @markus101 just thought I'd reach out to check the status on reviewing this branch.

markus101 commented on src/NzbDrone.Core/Download/Clients/qBittorrent/DigestAuthenticator.cs:
```diff
@@ -0,0 +1,22 @@
+
+namespace NzbDrone.Core.Download.Clients.QBittorrent
+{
+public class DigestAuthenticator : IAuthenticator
```
Indentation is wrong.

markus101 commented on src/NzbDrone.Core.Test/Download/DownloadClientTests/QBittorrentTests/QBittorrentFixture.cs:
```diff
@@ -0,0 +1,291 @@
+            id.Should().NotBeNullOrEmpty();
+        }
+
+        // Proxy.GetTorrents does not return original url. So item has to be found via magnet url.
```
This comment seems out of place since this isn't testing GetTorrents.

markus101 commented on src/NzbDrone.Core/Download/Clients/qBittorrent/QBittorrent.cs:
```diff
@@ -0,0 +1,252 @@
+        private OsPath GetOutputPath(QBittorrentTorrent torrent)
+        {
+            // is save_path supported by /query/torrents?
+            if (torrent.SavePath != null)
```
Is this the new property from your PR? Is there a new minimum API version that we can enforce and forego the block below? I'm quite happy forcing a minimum version for a new release.

cbodley replied:
Yes, they bumped to the API version to 5. I'll add a check for that in QBittorrent::TestConnection.

markus101 commented:
@cbodley very sorry about missing your comment, apparently I read the notification when travelling and didn't look at the PR again. Thanks @JustMaier for bumping this.

From the QB side it looks like they still haven't merged the label PR, which is unfortunate, but I don't think its mandatory, more a nice to have to prevent Sonarr from trying to process something obscure. Has the SavePath change been published to a new build?

From my POV this looks pretty solid, let me know on the above and we'll look at getting it merged. Again, sorry for the super long delay.

cbodley commented:
@markus101 The SavePath property did make it into the 3.2.4 release. The labels PR is part of the 3.3.0 milestone.

Note that `QBittorrentSettingsValidator` currently requires a non-empty TvCategory. So if we want to support pre-3.3.0 clients, I'll have to:

1. allow an empty TvCategory setting,
2. provide a useful error message if they set TvCategory with a pre-3.3.0 client (API version < 6), and
3. make the label-related bits conditional on having a TvCategory

I'll look into that this weekend.

cbodley commented:
@markus101 Okay, we should do the right thing now for 3.2.4 and 3.3.0+ clients.

cbodley commented:
There are check failures, but I don't see any evidence that they were caused by this pull request. If there's any action needed on my part, please let me know!

markus101 commented:
A good chunk of them are due to changes we made to fix the tests on our new build agents, could you rebase the PR off of latest develop please.

cbodley commented:
@markus101 Okay, I rebased and the green light is back!

cbodley commented:
According to https://github.com/qbittorrent/qBittorrent/issues/3800#issuecomment-158794604, the qBittorrent v3.3.0 release is planned for this coming weekend.

cbodley commented:
@markus101 v3.3.0 is out, and I've finally tested the new setLabel command. It didn't work at first because we give it a hash in uppercase. I added a patch for that, and it's working nicely.

markus101 commented on src/NzbDrone.Core/Download/Clients/qBittorrent/QBittorrentProxy.cs:
```diff
@@ -0,0 +1,210 @@
+            {
+                _logger.Info("Authentication required, logging in.");
+
+                var logged_in = _logins.Get(settings.Username + settings.Password, () => Login(client, settings), _loginTimeout);
```
`loggedIn` instead of `logged_in`

markus101 commented on src/NzbDrone.Core/Download/Clients/qBittorrent/QBittorrent.cs:
```diff
@@ -0,0 +1,248 @@
+            foreach (var torrent in torrents)
+            {
+                var item = new DownloadClientItem();
+                item.DownloadId = torrent.Hash;
```
We need to convert the Hash to upper case so Sonarr can track it properly.

markus101 commented on src/NzbDrone.Core/Download/Clients/qBittorrent/QBittorrent.cs:
```diff
@@ -0,0 +1,248 @@
+                item.TotalSize = torrent.Size;
+                item.DownloadClient = Definition.Name;
+                item.RemainingSize = (long)(torrent.Size * (1.0 - torrent.Progress));
+                item.RemainingTime = TimeSpan.FromSeconds(torrent.Eta);
```
The item need a category set so Sonarr can track things it didn't grab

markus101 commented on src/NzbDrone.Core/Download/Clients/qBittorrent/QBittorrent.cs:
```diff
@@ -0,0 +1,248 @@
+                    // API version 6 introduced support for labels
+                    return new NzbDroneValidationFailure("Category", "Category is not supported")
+                    {
+                        DetailedDescription = "Labels are not supported until qBittorrent version 3.3.0. Please try again with an empty Category."
```
Maybe: `Labels are not supported until qBittorrent version 3.3.0. Upgrade or try again with an empty Category.`?

cbodley replied:
@markus101 I did make this change. But on second thought, do you think it's worth keeping the compatibility code at all?

markus101 commented on src/NzbDrone.Core/Download/Clients/qBittorrent/QBittorrentProxy.cs:
```diff
@@ -0,0 +1,210 @@
+                return false;
+            }
+
+            if (response.Content != "Ok.") // returns "Fails." on bad login
```
Is this localized at all?

cbodley replied:
No, it's either 'Ok.' or 'Fails.' according to https://github.com/qbittorrent/qBittorrent/blob/master/src/webui/webapplication.cpp#L180

markus101 commented on src/NzbDrone.Core/Download/Clients/qBittorrent/QBittorrentSettings.cs:
```diff
@@ -0,0 +1,55 @@
+        }
+    }
+
+    public class QBittorrentSettings : IProviderConfig
```
Support for HTTPS should be added, since thats available in the client.

cbodley replied:
Okay, I'll look into that.

cbodley replied:
@markus101 I pushed a patch that adds a UseSsl setting as the other clients do. I was able to generate a certificate/key, add them to qBittorrent's webui, and test the client to verify that it works.

markus101 commented on src/NzbDrone.Core/Download/Clients/qBittorrent/QBittorrent.cs:
```diff
@@ -0,0 +1,249 @@
+        {
+            _proxy.AddTorrentFromUrl(magnetLink, Settings);
+
+            if (Settings.TvCategory.Length > 0)
```
`if (Settings.TvCategory.IsNotNullOrEmpty())`

markus101 commented on src/NzbDrone.Core/Download/Clients/qBittorrent/QBittorrent.cs:
```diff
@@ -0,0 +1,249 @@
+        {
+            _proxy.AddTorrentFromFile(filename, fileContent, Settings);
+
+            if (Settings.TvCategory.Length > 0)
```
Same here.

markus101 commented:
I think there are just a couple more things to fix up before we get it squashed and merged. :+1:

cbodley commented:
@markus101 excellent, thanks for all your help with review. I'll take care of those last comments this evening.

I did have a question for you regarding support for pre-3.3.0 clients, but it was hidden under an outdated diff:

Do you think it's worth keeping the compatibility code at all? At this point, it's probably more of a maintenance burden than it's worth, but that's ultimately up to you folks.

markus101 commented on src/NzbDrone.Core/Download/Clients/qBittorrent/QBittorrent.cs:
```diff
@@ -0,0 +1,249 @@
+                        DetailedDescription = "Please upgrade to qBittorrent version 3.2.4 or higher."
+                    };
+                }
+                else if (version < 6 && Settings.TvCategory.Length > 0)
```
Use IsNotNullOrEmpty() here as well.

markus101 commented:
I'm totally okay with making the minimum 3.3.0, I saw that we effectively make the minimum 3.2.4 because of `save_path` and then 3.3.0 if they set a category.

We don't req

cbodley commented:
@markus101 Okay, I added that warning to recommend setting the category when it's supported. I also did some squashing and cleanup to remove history related to the save_path from /query/propertiesGeneral.

markus101 commented:
Awesome @cbodley I'll try to test it tonight and get back to you ASAP. Thanks for implementing this.

markus101 commented on src/NzbDrone.Core/Download/Clients/qBittorrent/QBittorrent.cs:
```diff
@@ -0,0 +1,263 @@
+            {
+                var item = new DownloadClientItem();
+                item.DownloadId = torrent.Hash.ToUpper();
+                item.Category = Settings.TvCategory;
```
This needs to use the label from the response.

markus101 commented:
The unit tests are failing now and the one issue with the label, but everything else looks okay. Once those are fixed up please squash the commits into one and we can get it merged.

cbodley commented:
@markus101 the label issue is fixed, and everything is squashed and rebased on develop

markus101 commented:
Thanks @cbodley

JustMaier commented:
Thanks @cbodley - I really look forward to having this in Sonarr.

rterblanche commented:
When can we expect this in the develop branch or is there a beta branch. Will be happy to test this.
